File: repo_radar/activity/ids.py
import re, secrets, uuid

ACTIVITY_ID_RE = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")
TOKEN_RE = re.compile(r"^[0-9a-f]{8}$")

def mint_activity_id() -> str:
    return str(uuid.uuid4())

def valid_activity_id(s) -> bool:
    return isinstance(s, str) and bool(ACTIVITY_ID_RE.fullmatch(s))

def mint_token() -> str:
    return secrets.token_hex(4)

def valid_token(s) -> bool:
    return isinstance(s, str) and bool(TOKEN_RE.fullmatch(s))

File: menubar/activity/ids.js
'use strict';
// Node mirror of repo_radar/activity/ids.py (Task 2.1). Regexes are copied verbatim from the
// Python source (anchored `^...$`, used with `.test()`) -- Python validates with `re.fullmatch`,
// which for these patterns is exactly equivalent to a JS `^...$` test (both reject a dangling
// trailing char, including a trailing newline; verified empirically against CPython).
const crypto = require('crypto');

const ACTIVITY_ID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;
const TOKEN_RE = /^[0-9a-f]{8}$/;

function mintActivityId() {
  return crypto.randomUUID();
}

function validActivityId(s) {
  return typeof s === 'string' && ACTIVITY_ID_RE.test(s);
}

function mintToken() {
  return crypto.randomBytes(4).toString('hex');
}

function validToken(s) {
  return typeof s === 'string' && TOKEN_RE.test(s);
}

module.exports = { validActivityId, mintActivityId, mintToken, validToken };
